aiosmtpd/tests/test_server.py

```diff
@@ -17,42 +17,83 @@ def in_wsl():
-        s1.start()
-        self.assertRaises(socket.error, s2.start)
+        server.command_size_limits["DATA"] = 1024
+        assert server.max_command_size_limit == 1024
+
+    def test_warn_authreq_notls(self):
+        with pytest.warns(UserWarning) as record:
+            Server(Sink(), auth_require_tls=False, auth_required=True)
+        assert len(record) == 1
+        assert (
+            record[0].message.args[0]
+            == "Requiring AUTH while not requiring TLS can lead to "
+            "security vulnerabilities!"
+        )
+
+
+class TestController:
+    @pytest.mark.skipif(in_wsl(), reason="WSL prevents socket collision")
+    def test_socket_error_dupe(self, plain_controller, client):
+        contr2 = Controller(
+            Sink(), hostname=Global.SrvAddr.host, port=Global.SrvAddr.port
+        )
+        try:
+            with pytest.raises(socket.error):
+                contr2.start()
+        finally:
+            contr2.stop()
+
+    @pytest.mark.skipif(in_wsl(), reason="WSL prevents socket collision")
+    def test_socket_error_default(self):
+        contr1 = Controller(Sink())
+        contr2 = Controller(Sink())
+        try:
+            with pytest.raises(socket.error):
+                contr1.start()
+                contr2.start()
+        finally:
+            contr2.stop()
+            contr1.stop()
 
     def test_server_attribute(self):
         controller = Controller(Sink())
-        self.assertIsNone(controller.server)
+        assert controller.server is None
         try:
             controller.start()
-            self.assertIsNotNone(controller.server)
+            assert controller.server is not None
         finally:
             controller.stop()
-            self.assertIsNone(controller.server)
+        assert controller.server is None
+
+    def test_enablesmtputf8_flag(self):
+        # Default is True
+        controller = Controller(Sink())
+        assert controller.server_kwargs["enable_SMTPUTF8"]
+        # Explicit set must be reflected in server_kwargs
+        controller = Controller(Sink(), enable_SMTPUTF8=True)
+        assert controller.server_kwargs["enable_SMTPUTF8"]
+        controller = Controller(Sink(), enable_SMTPUTF8=False)
+        assert not controller.server_kwargs["enable_SMTPUTF8"]
+        # Explicit set must override server_kwargs
+        kwargs = dict(enable_SMTPUTF8=False)
+        controller = Controller(Sink(), enable_SMTPUTF8=True, server_kwargs=kwargs)
+        assert controller.server_kwargs["enable_SMTPUTF8"]
+        kwargs = dict(enable_SMTPUTF8=True)
+        controller = Controller(Sink(), enable_SMTPUTF8=False, server_kwargs=kwargs)
+        assert not controller.server_kwargs["enable_SMTPUTF8"]
+        # Set through server_kwargs must not be overridden if no explicit set
+        kwargs = dict(enable_SMTPUTF8=False)
+        controller = Controller(Sink(), server_kwargs=kwargs)
+        assert not controller.server_kwargs["enable_SMTPUTF8"]
```

Comment:
This tests behavior of `enable_SMTPUTF8` parameter interaction with `server_kwargs` parameter.

The behavior has been documented in the `controller.rst` file.